src/kubernetes/client/config/kubeconfig.rs

```diff
@@ -0,0 +1,231 @@
+//! Everything related to building a kubernetes client configuration from a kubeconfig file.
+//! Not all features of a kubeconfig are currently supported,
+//! e.g. password and OIDC auth as well as inlined certificates and keys.
+
+use super::Config;
+use crate::tls::TlsOptions;
+use snafu::{OptionExt, ResultExt, Snafu};
+
+use crate::kubernetes::client::config::kubeconfig_types::*;
+use std::path::{Path, PathBuf};
+
+impl Config {
+    /// Prepares a config by reading a kubeconfig file from defined path
+    pub fn kubeconfig(path: &Path) -> Result<Self, Error> {
+        kubeconfig_reader(&std::fs::read_to_string(path).context(IO)?)
+    }
+}
+
+fn kubeconfig_reader(config: &str) -> Result<Config, Error> {
+    let kc: Kubeconfig = serde_yaml::from_str(config).context(Parse)?;
+
+    // resolve "current_context"
+    let current_context = &kc.current_context;
+    let current_context = &kc
+        .contexts
+        .iter()
+        .find(|c| &c.name == current_context)
+        .context(MissingRelation {
+            from: "current_context",
+            missing: "context",
+        })?
+        .context;
+    let current_cluster = &current_context.cluster;
+    let current_user = &current_context.user;
+
+    // resolve cluster
+    let cluster = &kc
+        .clusters
+        .iter()
+        .find(|c| &c.name == current_cluster)
+        .context(MissingRelation {
+            from: "context.cluster",
+            missing: "cluster",
+        })?
+        .cluster;
+    // resolve user
+    let user = &kc
+        .auth_infos
+        .iter()
+        .find(|a| &a.name == current_user)
+        .context(MissingRelation {
+            from: "context.user",
+            missing: "auth_info",
+        })?
+        .auth_info;
+
+    let base = cluster.server.parse().context(InvalidUrl)?;
+
+    // use inline token if defined, otherwise fallback to token_file
+    let token = match &user.token {
+        Some(t) => Some(t.clone()),
+        None => match &user.token_file {
+            Some(file) => Some(std::fs::read_to_string(PathBuf::from(&file)).context(Token)?),
+            None => None,
+        },
+    };
```

Comment:
It's very possible that I'm missing something, but ultimately resolving the type errors when converting this to combinators leaves me doing battle with the borrow checker. If you're not vehemently opposed to the styling of nested matches I'd personally opt to leave this as is.

Reply:
The specific issue is because of the calling `read_to_string` inside of a map callback where we need to handle the result. We can implicitly convert it to an option but doing so leaves us with an `Option<Option<String>>` where our config struct expects an `Option<String>` then of course just unwrapping that is a no go because the user sits behind a shared ref.